File: atcoder/abc/python/abc219/abc219_E.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

class dsu :
    def __init__(self,n=1) :
        self.n = n
        self.parentOrSize = [-1 for i in range(n)]
    def merge(self,a,b) :
        x = self.leader(a); y = self.leader(b)
        if x == y : return x
        if self.parentOrSize[y] < self.parentOrSize[x] : (x,y) = (y,x)
        self.parentOrSize[x] += self.parentOrSize[y]
        self.parentOrSize[y] = x
        return x
    def leader(self,a) :
        if self.parentOrSize[a] < 0 : return a
        ans = self.leader(self.parentOrSize[a])
        self.parentOrSize[a] = ans
        return ans
    def size(self,a) :
        l = self.leader(a)
        return -self.parentOrSize[l]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    A = []
    for _ in range(4) : A.append(gis())
    targbm = 0
    for i in range(4) :
        for j in range(4) : 
            if A[i][j] == 1 : targbm |= 1 << (4*i+j)
    ans = 0 
    for bm in range(1<<16) :
        if bm & targbm != targbm : continue
        uf = dsu(17)
        numsq,root = 0,-1
        for i in range(16) :
            if bm & (1<<i) : numsq += 1; root = i
            if i+4 < 16 and bool(bm & (1<<i)) == bool(bm & (1<<(i+4))) : uf.merge(i,i+4)
            if i%4 != 3 and bool(bm & (1<<i)) == bool(bm & (1<<(i+1))) : uf.merge(i,i+1)
            if i not in (5,6,9,10) and bm & (1<<i) == 0 : uf.merge(i,16)
        if numsq == uf.size(root) and 17-numsq == uf.size(16) : ans += 1
    print(ans)

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: abc/go/abc219/abc219_E/abc219_E.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"math/bits"
	"os"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
var rdr = newScanner(os.Stdin)

func gs() string { rdr.sc.Scan(); return rdr.sc.Text() }
func gi() int { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gis(n int) []int { res := make([]int,n); for i:=0;i<n;i++ { res[i] = gi() }; return res }

type Dsu struct { n int; parentOrSize []int }
func NewDsu(n int) *Dsu { buf := make([]int, n); for i := 0; i < n; i++ { buf[i] = -1 }; return &Dsu{n, buf} }
func (q *Dsu) Leader(a int) int {
	if q.parentOrSize[a] < 0 { return a }; ans := q.Leader(q.parentOrSize[a]); q.parentOrSize[a] = ans; return ans
}
func (q *Dsu) Merge(a int, b int) int {
	x := q.Leader(a); y := q.Leader(b); if x == y { return x }; if q.parentOrSize[y] < q.parentOrSize[x] { x, y = y, x }
	q.parentOrSize[x] += q.parentOrSize[y]; q.parentOrSize[y] = x; return x
}
func (q *Dsu) Size(a int) int { l := q.Leader(a); return -q.parentOrSize[l] }

func solve(A [][]int) int {
	targbm,root := 0,-1
	for i:=0;i<4;i++ { 
		for j:=0;j<4;j++ { 
			if A[i][j] == 1 { 
				targbm |= 1 << (4*i+j); root = 4*i+j
			} 
		}
	}
	ans := 0
	for bm := 0; bm < 1<<16; bm++ {
		if bm & targbm != targbm { continue }
		uf := NewDsu(17)
		numsq := bits.OnesCount(uint(bm))
		for i:=0;i<12;i++ { 
			if (bm & (1<<i) == 0) == (bm & (1<<(i+4)) == 0) {
				uf.Merge(i,i+4)
			}
		}
		for i:=0;i<16;i++ { 
			if i % 4 == 3 { continue }
			if (bm & (1<<i) == 0) == (bm & (1<<(i+1)) == 0) { 
				uf.Merge(i,i+1)
			}
		}
		for i:=0;i<16;i++ { 
			if i==5 || i==6 || i==9 || i==10 { continue }
			if bm & (1<<i) == 0 { uf.Merge(i,16) }
		}
		if uf.Size(root)==numsq && uf.Size(16)==16-numsq+1 { ans++ }
	}
	return ans
}

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
    if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	A := make([][]int,4)
	for i:=0;i<4;i++ { A[i] = gis(4) }
	ans := solve(A)
	fmt.Println(ans)
}